### PR TITLE
Feat/more kernels

### PR DESCRIPTION
**Pull request by db091756**
### PR Type

- Feature
- Tests

### Description

Adds exponential, rational quadratic, periodic, locally periodic and polynomial kernel functions. #614 needs review first.
### How Has This Been Tested?

Unit tests for all new kernels.
Also adds test that checks that the implemented gradient and divergence of each kernel is equal to the autodifferentiation.

### Does this PR introduce a breaking change?

(Write your answer here.)

### Screenshots

(Write your answer here.)

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: tests/unit/test_kernel.py [hunk 1]
```diff
@@ -202,22 +222,371 @@ def expected_divergence_x_grad_y(
         """Compute expected divergence of the kernel w.r.t ``x`` gradient ``y``."""
 
 
+class KernelCompositionTest(ABC, Generic[_Kernel]):
```
Comment (db091756) [inline tests/unit/test_kernel.py:225]: Wasn't sure how to make this test class Generic on _AdditiveKernel and _ProductKernel

Comment (db091756) [reply tests/unit/test_kernel.py:225]: Or if we want that

File: coreax/kernel.py [hunk 2]
```diff
@@ -622,14 +685,246 @@ def divergence_x_grad_y_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
...
+            -grad_factor * func_1**2 * func_2 * func_3 * func_4
+            - grad_factor * factor * func_1 * func_2**2 * func_4
+            + grad_factor * factor * func_1 * func_3**2 * func_4
+            - (output_factor * sub * func_1**2 * func_2**2 * func_3**2 * func_4)
+        )
+
+        return output_factor * (d * first_term + jnp.dot(second_term, sub))
+
+
+class LocallyPeriodicKernel(ProductKernel):
```
Comment (pc532627) [inline coreax/kernel.py:884]: In this class, we assume the output scale and length scale are the same for both the inner RBF & periodic kernels. Might it be desired by the user to set these separately?

**Review by pc532627 — CHANGES_REQUESTED** [review]
One functional suggestion (shared parameters for the locally periodic kernel?) the rest just test and style related.

File: coreax/kernel.py [hunk 3]
```diff
@@ -578,7 +578,70 @@ def grad_y_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
...
+
+    output_scale: float = 1.0
+    constant: float = 0.0
+    degree: int = 2
+
+    def __check_init__(self):
+        """Ensure degree is an integer greater than 1."""
+        min_degree = 2
+        if not isinstance(self.degree, int) or self.degree < min_degree:
+            raise ValueError("'degree' must be a positive integer greater than 1")
```
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernel.py:607]: Added test that covers this line

File: coreax/kernel.py
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernel.py:1]: I'm not totally certain how you even reach line 406 in the current code. Running `SquaredExponentialKernel().gramian_row_mean(jnp.array([]))` fails at line 400 and gives a ZeroDivisionError which I think is fairly self explanatory?

File: coreax/kernel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pc532627) [reply coreax/kernel.py:607]: When I ran a coverage assessment it was still not hit

File: coreax/kernel.py (hunk 2 above)
Comment (pc532627) [reply coreax/kernel.py:883]: I think the docstring parameter entry for squared exponential output scale needs updating, currently reads:

```
    :param squared_exponential_output_scale: SquaredExponential Kernel
        smoothing/bandwidth parameter, :math:`\lambda`
```

on line 903-904

**Review by pc532627 — CHANGES_REQUESTED** [review]
2 Outstanding points but both should be quick to address

File: coreax/kernel.py (hunk 2 above)
Comment (db091756) [reply coreax/kernel.py:883]: Good spot, fixed

File: coreax/kernel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (db091756) [reply coreax/kernel.py:607]: Added now! Thought this was for __pow__

**Review by pc532627 — APPROVED** [review]
